### ISSUE TITLE
Chromecast tests failing

### ISSUE THREAD

**Issue by joeyparrish** [shaka-project/shaka-player#4094]
Tests are failing on Chromecast more often than not, but it's not the tests themselves.  The failures imply that some JavaScript files didn't load somehow.

```
Chrome 90.0.4430.225 (Chromecast 1.56.500000) ERROR
  TypeError: Cannot read property 'installAll' of undefined
      at test/test/boot.js:407:18 <- test/test/boot.js:496:18
```

```
Chrome 90.0.4430.225 (Chromecast 1.56.500000) ERROR
  TypeError: Cannot set property 'assert' of undefined
      at failTestsOnFailedAssertions (test/test/boot.js:50:3 <- test/test/boot.js:62:23)
      at test/test/boot.js:390:3 <- test/test/boot.js:479:3
```

In these two cases, lib/polyfill/all.js and lib/debug/asserts.js didn't load, respectively.  It's not clear how this is possible.

**Comment by joeyparrish** [comment]
To debug this, I made these changes:

1. I modified `node_modules/google-closure-library/closure/goog/base.js` to debug when scripts start and finish loading
2. I added `console.log` calls between every top-level statement in `lib/polyfill/all.js`
3. I added `console.log` calls in `test/test/boot.js` right before `shaka.polyfill.installAll()` is called

In all cases, I saw that `lib/polyfill/all.js` started and finished loading before `shaka.polyfill.installAll()` was called.  However, in the failure cases, I saw no logs at all from `lib/polyfill/all.js`.

I was testing the assumption that there was a silent exception thrown in `all.js`, which would result in some of the logs being shown, but not others.  Seeing no logs at all is especially confusing.  I don't understand how a script can apparently load, but have no effect.

I then added code in `boot.js` to show all `<script>` tags, and the loader definitely added a script tag referring to `all.js`, and it doesn't contain attributes like `async` or `defer`.

Finally, I tried changing the setting `goog.ENABLE_CHROME_APP_SAFE_SCRIPT_LOADING` to use an alternate loading mechanism, but this resulted in the tests consistently failing, with scripts now finishing loading **after** `boot.js`.

**Comment by joeyparrish** [comment]
In `test/test/boot.js`, it looks like we are explicitly setting `goog.ENABLE_DEBUG_LOADER` to `false`, so that calls to `goog.require` elsewhere in the tests will not do anything.  The thought was that by that point, we should have loaded everything via the require statements in `shaka-player.uncompiled.js`.

However, changing that setting at runtime appears to be interfering with the loader.  The timing is such that it still works on faster devices, but fails on a Chromecast.

The history of this is long and confusing, and there are artifacts of several changes over time.

We use `goog.require`/`goog.provide` in some of the tests, because at one time, the compiler would let us skip them in some cases, but not others.  Eventually, we changed `build/check.py` to add `--jscomp_off=missingRequire` when checking type safety in the tests.  So now all use of `goog.require`/`goog.provide` could be removed in `test/`.

The setting of `ENABLE_DEBUG_LOADER` was meant to avoid loading code via `goog.require`, and without it, there were errors about some classes being defined twice.  In retrospect, that was caused by a bug in `karma.conf.js` where some files in `demo/` were loaded explicitly outside of the `goog.require` loader.

**Comment by joeyparrish** [comment]
By removing all `goog.require`/`goog.provide` in `test/`, removing the setting of `ENABLE_DEBUG_LOADER` in `test/test/boot.js`, and fixing the inclusion of `demo/` code in `karma.conf.js`, the problem is greatly reduced, but still present.

Before making changes, in 10 test runs, the tests failed 8 times.

After making these changes, in 10 test runs, the tests failed 1-3 times in various runs.

I then tried replacing `shaka-player.uncompiled.js` in `karma.conf.js` with wildcards of all library files.  This would bypass the Closure loader and load all source files explicitly as `<script>` tags at the Karma level.

This approach still results in failures, but strangely, they become completely consistent.  Every run fails, and always the same way.  For example, missing `shaka.polyfill`.  Adding `console.log()` to the file that defines it causes that error to go away, and another to pop up.  Again, the failures would be completely consistent, but with a different class.  Each class I add logs to stops causing failures, and another takes its place.

I now have **no idea** what's going on.  😭

**Comment by joeyparrish** [comment]
I never found an explanation for the thing I commented on last, where console logging seemed to make requirements start working.  For now, I will assume that was caused by something I did wrong in my experimentation.  But I believe I have a working solution now.

I started working on the theory that something was wrong with the provide/require system, either in its implementation (unlikely) or in our use of it in tests (more likely).  I couldn't figure out what was wrong, but I found an alternative that is working better.

First, I dropped the use of provide/require in our tests.  In my earlier work on this, I realized that we never fully committed to provide/require in the tests, and ultimately adopted `--jscomp_off=missingRequire` when type-checking the tests.  This simplified things such that no new sources would be added at runtime by the tests.

Then, instead of having Karma load a handful of demo sources explicitly in `karma.conf.js`, I had Karma include the demo's `uncompiled.js` file, which has all the `goog.require` statements for the demo.  This loads the demo code under test in the same way as the main library, purely through require statements.

At this point, all library and demo code was being loaded through `goog.require`, at the time Karma loads sources listed in `karma.conf.js`.

To ensure that the timing is consistent and everything was loaded in the right order, I then wrote some code in `test/test/boot.js` to load the test utilities and test files in two subsequent stages, after Karma had finished.  This meant the library and demo would be loaded in stage 1, along with dependencies and the bootstrapping code in `boot.js`.  Then in stage 2, `boot.js` would load all test utils and wait for them to completely load.  Then in stage 3, `boot.js` would load the tests and wait for them, too.  Finally, `boot.js` would start Jasmine.

With this setup, tests completed successfully on Chromecast 100 times in a row.  Before my changes, the tests failed on Chromecast in 8 out of 10 runs.  So I believe the success of these changes is not a coincidence.

Since I got to this solution by wandering, I'm going to try undoing some parts of it to see if every one of these changes is needed, or if it would be possible to solve this in a simpler way.
